core/src/main/java/org/infinispan/persistence/manager/PersistenceManagerImpl.java

```diff
@@ -375,1083 +380,674 @@ public boolean isPreloaded() {
+            this::acquireReadLock,
+            ignore -> {
+               checkStoreAvailability();
+               if (trace) {
+                  log.trace("Writing batch to stores");
+               }
+               return Flowable.fromIterable(stores.entrySet())
+                     .filter(entry -> shouldWrite(entry.getValue(), predicate, flags))
+                     // Let the rollback work in parallel across the stores
+                     .flatMapCompletable(entry -> Completable.fromCompletionStage(
+                           PersistenceManagerImpl.<K, V>storeForEntry(entry).bulkWrite(segmentCount, flowable)));
+            },
+            this::releaseReadLock
+      ).toCompletionStage(null);
    }
 
    @Override
-   public CompletionStage<Integer> size(IntSet segments) {
-      acquireReadLock();
-      try {
-         checkStoreAvailability();
-         for (CacheLoader l : loaders) {
-            StoreConfiguration storeConfiguration;
-            if (l instanceof SegmentedAdvancedLoadWriteStore &&
-                  ((storeConfiguration = getStoreConfig(l)) != null && storeConfiguration.segmented())) {
-               return supplyOnPersistenceExAndContinue(traceId -> {
-                  if (trace) {
-                     log.tracef("Continuing size operation for id %d", traceId);
-                  }
-                  return ((SegmentedAdvancedLoadWriteStore) l).size(segments);
-               }, "Retrieving size with segments for id %d");
-            }
-         }
-         if (trace) {
-            log.tracef("Calculating size of store via publisher for segments %s", segments);
-         }
-
-         return Flowable.fromPublisher(publishKeys(segments, null, AccessMode.BOTH))
-               .count()
-               .map(count -> {
-                  long longValue = count;
-                  if (longValue > Integer.MAX_VALUE) {
-                     return Integer.MAX_VALUE;
-                  }
-                  return (int) longValue;
-               })
-               .toCompletionStage();
-
-      } finally {
-         releaseReadLock();
-      }
+   public CompletionStage<Void> deleteBatchFromAllNonTxStores(Iterable<Object> keys, Predicate<? super StoreConfiguration> predicate, long flags) {
+      Flowable<NonBlockingStore.SegmentedPublisher<Object>> flowable = Flowable.fromIterable(keys)
+            .groupBy(keyPartitioner::getSegment)
+            .map(SegmentPublisherWrapper::new);
+      return Completable.using(
+            this::acquireReadLock,
+            ignore -> {
+               checkStoreAvailability();
+               if (trace) {
+                  log.trace("Deleting batch of entries from stores");
+               }
+               return Flowable.fromIterable(stores.entrySet())
+                     .filter(entry ->
+                           !entry.getValue().characteristics.contains(Characteristic.READ_ONLY)
+                                 && predicate.test(entry.getValue().config))
```

Comment:
`shouldWrite`?

Reply:
Technically we are doing the same thing as the old PersistenceManagerImpl as the flags are only looked at for the single value write. I think I will leave it this way and create a JIRA. Because I would really like to figure out some other way to handle this rolling upgrade case, but I don't know how yet.

Reply:
I created https://issues.redhat.com/browse/ISPN-11859